# Coffee Shop Sample Data

## About dataset
This sample data module contains representative retail data from ficional coffee chain. Source: IBM (https://www.kaggle.com/datasets/ylchang/coffee-shop-sample-data-1113/data)

## Business Problem
A coffee shop wants to segment their customers and define marketing strategies according to these segments. For this purpose, behaviors of customers will be defined and groups will be formed according to clusters in behaviors.

## Import Lib, Files

In [2]:
import pandas as pd
import datetime as dt

In [3]:
# read csv data
df_ = pd.read_csv('D:\dataset\Coffee-shop-sample-data\\april 2019 sales reciepts.csv')
df = df_.copy()
df

transaction_id transaction_date transaction_time  sales_outlet_id  \
0                   7       2019-04-01         12:04:43                3   
1                  11       2019-04-01         15:54:39                3   
2                  19       2019-04-01         14:34:59                3   
3                  32       2019-04-01         16:06:04                3   
4                  33       2019-04-01         19:18:37                3   
...               ...              ...              ...              ...   
49889             753       2019-04-29         16:51:58                8   
49890             756       2019-04-29         16:51:14                8   
49891             759       2019-04-29         11:17:36                8   
49892             763       2019-04-29         15:45:52                8   
49893             763       2019-04-29         15:45:52                8   

       staff_id  customer_id instore_yn  order  line_item_id  product_id  \
0            12          558          N      1             1          52   
1            17          781          N      1             1          27   
2            17          788          Y      1             1          46   
3            12          683          N      1             1          23   
4            17           99          Y      1             1          34   
...         ...          ...        ...    ...           ...         ...   
49889        42            0          N      1             1          30   
49890        42         8412          Y      1             1          25   
49891        15            0          Y      1             1          31   
49892        45         8030          N      1             1          44   
49893        45         8030          N      1             5          75   

       quantity  line_item_amount  unit_price promo_item_yn  
0             1              2.50        2.50             N  
1             2              7.00        3.50             N  
2             2              5.00        2.50             N  
3             2              5.00        2.50             N  
4             1              2.45        2.45             N  
...         ...               ...         ...           ...  
49889         1              3.00        3.00             N  
49890         2              4.40        2.20             N  
49891         1              2.20        2.20             N  
49892         2              5.00        2.50             N  
49893         1              3.50        3.50             N  

[49894 rows x 14 columns]

## Data Understanding

In [4]:
# assess
df.info()

# wrong datatipes for few columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49894 entries, 0 to 49893
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   transaction_id    49894 non-null  int64  
 1   transaction_date  49894 non-null  object 
 2   transaction_time  49894 non-null  object 
 3   sales_outlet_id   49894 non-null  int64  
 4   staff_id          49894 non-null  int64  
 5   customer_id       49894 non-null  int64  
 6   instore_yn        49894 non-null  object 
 7   order             49894 non-null  int64  
 8   line_item_id      49894 non-null  int64  
 9   product_id        49894 non-null  int64  
 10  quantity          49894 non-null  int64  
 11  line_item_amount  49894 non-null  float64
 12  unit_price        49894 non-null  float64
 13  promo_item_yn     49894 non-null  object 
dtypes: float64(2), int64(8), object(4)
memory usage: 5.3+ MB


'transaction_date' is in object type, so it need to be changed to datetime

In [5]:
# change transaction_date into datetime datatype
df["transaction_date"] = df["transaction_date"].apply(pd.to_datetime)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49894 entries, 0 to 49893
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    49894 non-null  int64         
 1   transaction_date  49894 non-null  datetime64[ns]
 2   transaction_time  49894 non-null  object        
 3   sales_outlet_id   49894 non-null  int64         
 4   staff_id          49894 non-null  int64         
 5   customer_id       49894 non-null  int64         
 6   instore_yn        49894 non-null  object        
 7   order             49894 non-null  int64         
 8   line_item_id      49894 non-null  int64         
 9   product_id        49894 non-null  int64         
 10  quantity          49894 non-null  int64         
 11  line_item_amount  49894 non-null  float64       
 12  unit_price        49894 non-null  float64       
 13  promo_item_yn     49894 non-null  object        
dtypes: datetime64[ns](1), 

In [6]:
df.describe()

transaction_id  sales_outlet_id      staff_id   customer_id  \
count    49894.000000     49894.000000  49894.000000  49894.000000   
mean       869.056059         5.351846     25.359582   2282.324468   
std        857.863149         2.074796     12.466490   3240.551757   
min          1.000000         3.000000      6.000000      0.000000   
25%        223.000000         3.000000     15.000000      0.000000   
50%        481.000000         5.000000     26.000000      0.000000   
75%       1401.000000         8.000000     41.000000   5412.000000   
max       4203.000000         8.000000     45.000000   8501.000000   

              order  line_item_id    product_id      quantity  \
count  49894.000000  49894.000000  49894.000000  49894.000000   
mean       1.173428      1.631860     47.878983      1.438209   
std        1.025445      1.412881     17.928355      0.543039   
min        1.000000      1.000000      1.000000      1.000000   
25%        1.000000      1.000000     33.000000      1.000000   
50%        1.000000      1.000000     47.000000      1.000000   
75%        1.000000      1.000000     60.000000      2.000000   
max        9.000000     12.000000     87.000000      8.000000   

       line_item_amount    unit_price  
count      49894.000000  49894.000000  
mean           4.682646      3.384645  
std            4.436668      2.682545  
min            0.000000      0.800000  
25%            3.000000      2.500000  
50%            3.750000      3.000000  
75%            6.000000      3.750000  
max          360.000000     45.000000

50% of the transaction made by customer with id '0'

In [7]:
df.nunique()

transaction_id       4203
transaction_date       29
transaction_time    26074
sales_outlet_id         3
staff_id               25
customer_id          2248
instore_yn              3
order                   9
line_item_id           11
product_id             80
quantity                6
line_item_amount       77
unit_price             41
promo_item_yn           2
dtype: int64

In [8]:
df.groupby("customer_id").agg({"transaction_id" : "count"}).sort_values(by="transaction_id", ascending=False).head(15)

transaction_id
customer_id                
0                     25033
3                        31
8285                     31
548                      31
8341                     29
8009                     28
8158                     28
8118                     28
8036                     28
8375                     27
8395                     27
8446                     27
8101                     27
8059                     27
8384                     27

It seems imposible to made 25,033 transactions in a month

In [54]:
# examine customer_id 0
df[df["customer_id"] == 0].describe().T

count        mean         std  min    25%     50%  \
transaction_id    25033.0  846.087684  872.662668  1.0  220.0  459.00   
sales_outlet_id   25033.0    5.462749    2.112900  3.0    3.0    5.00   
staff_id          25033.0   26.008908   12.722100  6.0   15.0   26.00   
customer_id       25033.0    0.000000    0.000000  0.0    0.0    0.00   
order             25033.0    1.217513    1.151455  1.0    1.0    1.00   
line_item_id      25033.0    1.645188    1.409440  1.0    1.0    1.00   
product_id        25033.0   48.142412   18.326104  1.0   33.0   47.00   
quantity          25033.0    1.440898    0.557597  1.0    1.0    1.00   
line_item_amount  25033.0    4.766217    5.166521  0.0    3.0    3.75   
unit_price        25033.0    3.455893    3.024623  0.8    2.5    3.00   

                      75%     max  
transaction_id    1311.00  4202.0  
sales_outlet_id      8.00     8.0  
staff_id            42.00    45.0  
customer_id          0.00     0.0  
order                1.00     9.0  
line_item_id         1.00    12.0  
product_id          61.00    87.0  
quantity             2.00     8.0  
line_item_amount     6.00   360.0  
unit_price           3.75    45.0

In [56]:
# compare with customer id '0'
df[df["customer_id"] != 0].describe().T

count         mean          std  min    25%      50%  \
transaction_id    24861.0   892.183339   842.083445  1.0  226.0   519.00   
sales_outlet_id   24861.0     5.240175     2.029637  3.0    3.0     5.00   
staff_id          24861.0    24.705764    12.169001  6.0   15.0    20.00   
customer_id       24861.0  4580.439122  3247.865616  1.0  584.0  5417.00   
order             24861.0     1.129037     0.878318  1.0    1.0     1.00   
line_item_id      24861.0     1.618439     1.416239  1.0    1.0     1.00   
product_id        24861.0    47.613732    17.515098  1.0   33.0    46.00   
quantity          24861.0     1.435501     0.527972  1.0    1.0     1.00   
line_item_amount  24861.0     4.598497     3.551520  0.0    3.0     3.75   
unit_price        24861.0     3.312905     2.284791  0.8    2.5     3.00   

                      75%     max  
transaction_id    1485.00  4203.0  
sales_outlet_id      8.00     8.0  
staff_id            30.00    45.0  
customer_id       8099.00  8501.0  
order                1.00     9.0  
line_item_id         1.00    10.0  
product_id          59.00    87.0  
quantity             2.00     8.0  
line_item_amount     6.00   360.0  
unit_price           3.75    45.0

We cannot determine why customer ID '0' has 25,033 total transactions, but it is possible that the system has mistakenly assigned transactions from other customer IDs to '0'.
Therefore we will remove all customer id '0' 

In [50]:
# remove unusual variables
df1 = df[df["customer_id"] != 0]
df1

transaction_id transaction_date transaction_time  sales_outlet_id  \
0                   7       2019-04-01         12:04:43                3   
1                  11       2019-04-01         15:54:39                3   
2                  19       2019-04-01         14:34:59                3   
3                  32       2019-04-01         16:06:04                3   
4                  33       2019-04-01         19:18:37                3   
...               ...              ...              ...              ...   
49882             740       2019-04-29         08:16:20                8   
49888             752       2019-04-29         16:36:24                8   
49890             756       2019-04-29         16:51:14                8   
49892             763       2019-04-29         15:45:52                8   
49893             763       2019-04-29         15:45:52                8   

       staff_id  customer_id instore_yn  order  line_item_id  product_id  \
0            12          558          N      1             1          52   
1            17          781          N      1             1          27   
2            17          788          Y      1             1          46   
3            12          683          N      1             1          23   
4            17           99          Y      1             1          34   
...         ...          ...        ...    ...           ...         ...   
49882        15         8026          Y      1             1          41   
49888        42         8401          N      1             1          87   
49890        42         8412          Y      1             1          25   
49892        45         8030          N      1             1          44   
49893        45         8030          N      1             5          75   

       quantity  line_item_amount  unit_price promo_item_yn  
0             1              2.50        2.50             N  
1             2              7.00        3.50             N  
2             2              5.00        2.50             N  
3             2              5.00        2.50             N  
4             1              2.45        2.45             N  
...         ...               ...         ...           ...  
49882         1              4.25        4.25             N  
49888         1              2.10        2.10             Y  
49890         2              4.40        2.20             N  
49892         2              5.00        2.50             N  
49893         1              3.50        3.50             N  

[24861 rows x 14 columns]

In [63]:
a = df1.groupby(['sales_outlet_id']).agg({'quantity': 'mean',
                                'line_item_amount': 'mean',
                                'transaction_id': 'count'})
a.columns = ['average number of products purchased', 'average spend', 'total transaction']
a

average number of products purchased  average spend  \
sales_outlet_id                                                        
3                                            1.401024       4.486686   
5                                            1.479471       4.673514   
8                                            1.425185       4.639885   

                 total transaction  
sales_outlet_id                     
3                             8593  
5                             8549  
8                             7719

## Calculating RFM

In [44]:
# Check the Maximum transaction date 
print(df1['transaction_date'].max())

# For calculationg recency, will use 2019-05-01
max_date = dt.datetime(2019, 5, 1)
print(max_date)


2019-04-29 00:00:00
2019-05-01 00:00:00


2019-05-01 will be used as the maximum date to calculate recency

In [45]:
rfm = df1.groupby("customer_id").agg({"transaction_date" : lambda trx_date: (max_date - trx_date.max()).days,
                                     "transaction_id" : lambda trx_id: trx_id.nunique(),
                                     "line_item_amount" : lambda item_amount: item_amount.sum()})

# Name the columns as recency, frequency, and monetary
rfm.columns = ['recency', 'frequency', 'monetary']

print(rfm.head(15))
print(rfm.describe().T)

             recency  frequency  monetary
customer_id                              
1                  2          6     29.20
2                  5         16     90.35
3                  7         13    188.90
4                  5          7     28.75
5                  2          6     23.50
6                 10          9     44.05
7                  3          9     54.50
8                  5         10     61.20
9                  5          9     54.25
10                 2         10     46.90
11                 3          8     37.15
12                 4          7     37.00
13                 9          5     25.95
14                 2         10     41.60
15                10         11     65.50
            count       mean        std   min   25%   50%     75%     max
recency    2247.0   6.073431   5.248707  2.00   3.0   4.0   7.000   30.00
frequency  2247.0   8.734312   3.749882  1.00   6.0   8.0  11.000   25.00
monetary   2247.0  50.878162  25.645834  2.45  33.2  47.0  65.22

In [46]:
# Rank the RFM Score
rfm['recency_score'] = pd.qcut(rfm['recency'].rank(method='first'), 5, labels=[5, 4, 3, 2, 1])
# rank method helps by assigning the first comer number to first label due to repetitive numbers
rfm['frequency_score'] = pd.qcut(rfm['frequency'], 5, labels=[1, 2, 3, 4, 5])
rfm['monetary_score'] = pd.qcut(rfm['monetary'], 5, labels=[1, 2, 3, 4, 5])

# Define rfm_score
rfm['RFM_score'] = (rfm['recency_score'].astype(str) + rfm['frequency_score'].astype(str))
# monetary score is necessary for observation, but it is not used in forming the RFM score

RFM Graph reference: https://medium.com/@denizcansuturan/rfm-analysis-3930a9a5238 


In [47]:
# Segmentation Maps based on the RFM Graph
segm_map = {
    r'[1-2][1-2]': 'hibernating',
    r'[1-2][3-4]': 'at_risk',
    r'[1-2]5': 'cant_loose',
    r'3[1-2]': 'about_to_sleep',
    r'33': 'need_attention',
    r'[3-4][4-5]': 'loyal_customers',
    r'41': 'promising',
    r'51': 'new_customers',
    r'[4-5][2-3]': 'potential_loyalists',
    r'5[4-5]': 'champions'   
}

# the R-F table is coded using regex
rfm['segment'] = rfm['RFM_score'].replace(segm_map, regex=True)

In [48]:
rfm.head()

recency  frequency  monetary recency_score frequency_score  \
customer_id                                                               
1                  2          6     29.20             5               1   
2                  5         16     90.35             3               5   
3                  7         13    188.90             2               5   
4                  5          7     28.75             3               2   
5                  2          6     23.50             5               1   

            monetary_score RFM_score          segment  
customer_id                                            
1                        1        51    new_customers  
2                        5        35  loyal_customers  
3                        5        25       cant_loose  
4                        1        32   about_to_sleep  
5                        1        51    new_customers

In [49]:
segment_analysis = rfm.groupby('segment').agg({'recency': 'mean',
                                               'frequency': 'mean',
                                               'monetary': 'mean',
                                               'segment': 'count'})
segment_analysis.columns = ['r_mean', 'f_mean', 'm_mean', 'segm_count']
segment_analysis

r_mean     f_mean     m_mean  segm_count
segment                                                         
about_to_sleep        4.419580   5.545455  35.060490         143
at_risk               8.853026   9.757925  55.824553         347
cant_loose            7.242718  15.058252  88.756796         103
champions             2.000000  12.036496  70.410073         137
hibernating          12.973274   5.118040  29.913920         449
loyal_customers       3.267640  13.072993  75.271411         411
need_attention        4.317073   8.548780  51.320366          82
new_customers         2.000000   5.157143  28.993929         140
potential_loyalists   2.448795   7.954819  46.779217         332
promising             2.990291   4.970874  28.973883         103

In [51]:
# save to csv
#rfm.to_csv("rfm.csv")

In [52]:
#df1.to_csv("clean_sales_data_april.csv")